core/src/main/scala/kafka/admin/AdminUtils.scala

```diff
@@ -235,13 +361,30 @@ object AdminUtils extends Logging {
   def topicExists(zkUtils: ZkUtils, topic: String): Boolean =
     zkUtils.zkClient.exists(getTopicPath(topic))
 
+  def getBrokersAndRackInfo(zkUtils: ZkUtils, rackAwareMode: RackAwareMode = RackAwareMode.Enforced, brokerList: Seq[Int] = Seq()) = {
+    val allBrokers = zkUtils.getAllBrokersInCluster()
+    val brokers = if (brokerList.isEmpty) allBrokers else allBrokers.filter(b => brokerList.contains(b.id))
+    val brokersWithRack = brokers.filter(_.rack.nonEmpty)
+    if (rackAwareMode == RackAwareMode.Enforced && brokersWithRack.size > 0 && brokersWithRack.size < brokers.size) {
+      throw new AdminOperationException("Not all brokers have rack information. Add --disable-rack-aware in command line" +
+        " to make replica assignment without rack information.")
+    }
+    val brokerRackMap: Map[Int, String] = rackAwareMode match {
+      case RackAwareMode.Disabled => Map()
+      case RackAwareMode.Safe if (brokersWithRack.size < brokers.size) => Map()
+      case _ => brokersWithRack.map(broker => (broker.id -> broker.rack.get)).toMap
+    }
+    (brokers.map(_.id).sorted, brokerRackMap)
+  }
+
   def createTopic(zkUtils: ZkUtils,
                   topic: String,
                   partitions: Int,
                   replicationFactor: Int,
-                  topicConfig: Properties = new Properties) {
-    val brokerList = zkUtils.getSortedBrokerList()
-    val replicaAssignment = AdminUtils.assignReplicasToBrokers(brokerList, partitions, replicationFactor)
+                  topicConfig: Properties = new Properties,
+                  rackAwareMode: RackAwareMode = RackAwareMode.Enforced) {
```

Comment:
"Safe" is only used in auto topic creation. In command line tools, we would like to be strict about using rack (to catch mis-configured rack) unless the user wants to disable it. This was discussed in KIP process.

The reason to make this argument optional is that in most cases, user would supply rack for all brokers or no rack for any broker which can be handled automatically in "Enforced" mode. Then createTopic can remain the same signature so that caller of this method does not need to be concerned about rack aware.